src/common/hist_util.cc

```diff
@@ -412,25 +451,45 @@ void GHistIndexMatrix::Init(DMatrix* p_fmat, int max_num_bins) {
       }
     }
 
-    index.resize(row_ptr[rbegin + batch.Size()]);
-
-    CHECK_GT(cut.Values().size(), 0U);
+//    index.resize(row_ptr[rbegin + batch.Size()]);
+    const size_t n_disps = cut.Ptrs().size() - 1;
+    if ((max_num_bins - 1 <= static_cast<int>(std::numeric_limits<uint8_t>::max())) && isDense) {
+      index.setBinBound(POWER_OF_TWO_8);
+      index.resize((sizeof(uint8_t)) * row_ptr[rbegin + batch.Size()], n_disps);
+    } else if ((max_num_bins - 1 > static_cast<int>(std::numeric_limits<uint8_t>::max())  &&
+            max_num_bins - 1<= static_cast<int>(std::numeric_limits<uint16_t>::max())) && isDense) {
+      index.setBinBound(POWER_OF_TWO_16);
+      index.resize((sizeof(uint16_t)) * row_ptr[rbegin + batch.Size()], n_disps);
+    } else {
+      index.setBinBound(POWER_OF_TWO_32);
+      index.resize((sizeof(uint32_t)) * row_ptr[rbegin + batch.Size()], n_disps);
+    }
 
-    #pragma omp parallel for num_threads(batch_threads) schedule(static)
-    for (omp_ulong i = 0; i < batch.Size(); ++i) { // NOLINT(*)
-      const int tid = omp_get_thread_num();
-      size_t ibegin = row_ptr[rbegin + i];
-      size_t iend = row_ptr[rbegin + i + 1];
-      SparsePage::Inst inst = batch[i];
 
-      CHECK_EQ(ibegin + inst.size(), iend);
-      for (bst_uint j = 0; j < inst.size(); ++j) {
-        uint32_t idx = cut.SearchBin(inst[j]);
+    CHECK_GT(cut.Values().size(), 0U);
 
-        index[ibegin + j] = idx;
-        ++hit_count_tloc_[tid * nbins + idx];
+    uint32_t* disps = index.disp();
+    index.setDispSize(n_disps);
+    if ((max_num_bins <= 65536) && isDense) {
```

Comment:
1) Where is case for <= 256?
2) Again, 65536 is magic constant, use std::numeric_limits

Reply:
It merged case. Only for max_bin > std::numeric_limit<uint16_t>() we should set displaces to 0.

Reply:
So, you have 2 branches:
- If max_bin <= std::numeric_limit<uint16_t>() you use offsets for bins
- else you don't use them or set them to zeros, and bin indexes contain features offsets already

Do I understand this correctly? If so - I dont see the reason for such behavior.

Reply:
Displaces were added only for dense data sets, 
as for sparse we should keep feature ids, for each element (uint32_t), and in this case there is no reasons to use reduced bin indexes.